### PR TITLE
SURT dataset

### PR DESCRIPTION
**Pull request by desh2608**
This dataset is for multi-talker ASR, e.g., using the SURT models (see this paper: https://arxiv.org/abs/2109.08555). It can also be modified for other E2E multi-talker ASR tasks e.g. SOT training (like in ESPNet).

@pzelasko This is finally ready for review!

### CODE REVIEW

File: lhotse/cut/set.py [hunk 1]
```diff
@@ -620,7 +620,6 @@ def to_shar(
                 warn_unused_fields=warn_unused_fields,
                 include_cuts=include_cuts,
                 shard_suffix=None,
-                verbose=verbose,
```
Comment (pzelasko) [inline lhotse/cut/set.py:623]: should this be removed?

File: lhotse/cut/set.py [hunk 2]
```diff
@@ -641,7 +640,6 @@ def to_shar(
                         warn_unused_fields=warn_unused_fields,
                         include_cuts=True,
                         shard_suffix=f".{idx:06d}",
-                        verbose=False,
```
Comment (pzelasko) [inline lhotse/cut/set.py:644]: should this be removed?

File: lhotse/cut/set.py [hunk 3]
```diff
@@ -3418,12 +3416,9 @@ def _export_to_shar_single(
     warn_unused_fields: bool,
     include_cuts: bool,
     shard_suffix: Optional[str],
-    verbose: bool,
```
Comment (pzelasko) [inline lhotse/cut/set.py:3421]: should this be removed?

File: lhotse/cut/set.py [hunk 4]
```diff
@@ -3418,12 +3416,9 @@ def _export_to_shar_single(
...
 ) -> Dict[str, List[str]]:
     from lhotse.shar import SharWriter
 
-    pbar = tqdm(desc="Exporting to SHAR", disable=not verbose)
```
Comment (pzelasko) [inline lhotse/cut/set.py:3425]: should this be removed?

File: lhotse/cut/set.py [hunk 5]
```diff
@@ -3434,7 +3429,5 @@ def _export_to_shar_single(
     ) as writer:
         for cut in cuts:
             writer.write(cut)
-            pbar.update()
```
Comment (pzelasko) [inline lhotse/cut/set.py:3437]: should this be removed?

File: lhotse/dataset/surt.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    Each item in this dataset is a dict of:
+
+    .. code-block::
+
+        {
+            'inputs': float tensor with shape determined by :attr:`input_strategy`:
+                      - single-channel:
+                        - features: (B, T, F)
+                        - audio: (B, T)
+                      - multi-channel: currently not supported
```
Comment (pzelasko) [inline lhotse/dataset/surt.py:62]: is it really not supported? the rest of the docs suggests otherwise

File: lhotse/cut/set.py [hunk 7]
```diff
@@ -2640,7 +2638,7 @@ def mix(
         )
         snr = None
 
-    if reference_cut.num_features is not None and mixed_in_cut.num_features is not None:
```
Comment (pzelasko) [inline lhotse/cut/set.py:2643]: is this a bugfix or accidental change?

File: lhotse/dataset/surt.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+from lhotse.workarounds import Hdf5MemoryIssueFix
+
+
+class K2SurtDataset(torch.utils.data.Dataset):
+    """
+    The PyTorch Dataset for the multi-talker ASR task using k2 library.
+    We support the modeling framework known as Streaming Unmixing and Recognition
+    Transducer (SURT), as described in [1] and [2], but this dataset can also be
+    used for other multi-talker ASR approaches such as MT-RNNT [3] and SOT [4].
+    See icefall recipe for usage.
```
Comment (pzelasko) [inline lhotse/dataset/surt.py:21]: you might want to link the recipe here -- otherwise it's a nice doc, very helpful :)

**Review by pzelasko — COMMENTED** [review]
Looks good to me! I left a few questions, as I'm not sure if every change is intentional.

File: lhotse/dataset/surt.py (hunk 8 above)
Comment (desh2608) [reply lhotse/dataset/surt.py:21]: Yeah, I am cleaning up the icefall recipe to make a PR. I will link it here once it is done.

File: lhotse/dataset/surt.py (hunk 6 above)
Comment (desh2608) [reply lhotse/dataset/surt.py:62]: Here by "channel" we are referring to the audio channel (i.e. number of microphones) in the input recording. In the rest of the code, "channel" actually refers to an output channel on which a transcription is done. For example, overlapping speech may be transcribed simultaneously on 2 output channels.

**Review by desh2608 — COMMENTED** [review]
I have removed the accidental changes. Thanks for pointing out!

**Comment by desh2608** [comment]
I have added unit tests and linked the icefall PR. @pzelasko could you approve when you have time?

**Review by pzelasko — DISMISSED** [review]
LGTM!

**Comment by desh2608** [comment]
> LGTM!

Sorry I had missed adding the test file. Added it now!

**Review by pzelasko — APPROVED** [review]

**Pull request merged**